In [196]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
import re

In [238]:
os.getcwd()

'C:\\Users\\sunilchawla\\Desktop\\ass2\\code'

In [239]:
path = "../data_clean/"

In [240]:
df_train = pd.read_csv(path+"pos_neg.csv",sep = ",")
df_test = pd.read_csv(path+"pos_neg_test.csv",sep = ",")

In [241]:
df_train.drop(['Unnamed: 0'], axis = 1, inplace = True)
df_test.drop(['Unnamed: 0'], axis = 1, inplace = True)

In [242]:
df_train.head()

movieID                                             review  rate  label  \
0        0  story of a man who has unnatural feelings for ...     3      0   
1    10000  airport '77 starts as a brand new luxury 747 p...     4      0   
2    10001  this film lacked something i could not put my ...     4      0   
3    10002  sorry everyone,,, i know this is supposed to b...     1      0   
4    10003  when i was little my parents took me along to ...     1      0   

                                               words  \
0  ['story', 'man', 'unnatural', 'feeling', 'pig'...   
1  ['airport', "'77", 'start', 'brand', 'new', 'l...   
2  ['film', 'lacked', 'something', 'could', 'not'...   
3  ['sorry', 'everyone', 'know', 'supposed', '``'...   
4  ['little', 'parent', 'took', 'along', 'theater...   

                                          words_sent  
0  story man unnatural feeling pig start opening ...  
1  airport '77 start brand new luxury 747 plane l...  
2  film lacked something could not put finger fir...  
3  sorry everyone know supposed  art  film wow ha...  
4  little parent took along theater see interior ...

In [243]:
df_test.head()

movieID                                             review  rate  label  \
0        0  once again mr. costner has dragged out a movie...     2      0   
1    10000  this is an example of why the majority of acti...     4      0   
2    10001  first of all i hate those moronic rappers, who...     1      0   
3    10002  not even the beatles could write songs everyon...     3      0   
4    10003  brass pictures (movies is not a fitting word f...     3      0   

                                               words  \
0  ['again', 'mr.', 'costner', 'dragged', 'movie'...   
1  ['example', 'majority', 'action', 'film', 'gen...   
2  ['first', 'all', 'hate', 'moronic', 'rapper', ...   
3  ['not', 'even', 'beatles', 'could', 'write', '...   
4  ['brass', 'picture', 'movie', 'not', 'fitting'...   

                                          words_sent  
0  again mr. costner dragged movie far longer nec...  
1  example majority action film generic boring re...  
2  first all hate moronic rapper could'nt act gun...  
3  not even beatles could write song everyone lik...  
4  brass picture movie not fitting word really so...

# <font color="red">preprocessed Data</font>

In [244]:
X_train=df_train['words_sent'].str.lower()
X_test=df_test['words_sent'].str.lower()
Y_train = df_train['label']
Y_test = df_test['label']

# <font color="red">Hyperparameter tuning with preprocessed Data</font>

In [231]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_vect = TfidfVectorizer(ngram_range=(1, 3),min_df=3,max_df=0.56)
train_document = tfidf_vect.fit_transform(X_train)
test_document = tfidf_vect.transform(X_test)

In [72]:
from sklearn.model_selection import GridSearchCV

In [23]:
from sklearn.naive_bayes import MultinomialNB
param_grid = {'alpha': [0.01, 0.1, 1, 10]}
grid_search = GridSearchCV(MultinomialNB(),param_grid, cv = 3)
grid_search.fit(train_document, Y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'alpha': [0.01, 0.1, 1, 10]}, pre_dispatch='2*n_jobs',
       refit=True, return_train_score='warn', scoring=None, verbose=0)

In [234]:
from sklearn.naive_bayes import MultinomialNB
clf = BernoulliNB().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.4f}".format(clf.score(test_document, Y_test)))

Training set score: 0.97
Test set score: 0.8755


In [263]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_vect = TfidfVectorizer(ngram_range=(1, 3),min_df=2,max_df=0.56)
train_document = tfidf_vect.fit_transform(X_train)
test_document = tfidf_vect.transform(X_test)

In [25]:
param_grid = {'C': [0.01, 0.1, 1, 10,100]}
from sklearn.linear_model import LogisticRegression
grid_search = GridSearchCV(LogisticRegression(),param_grid, cv = 3)
grid_search.fit(train_document, Y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'C': [0.01, 0.1, 1, 10, 100]}, pre_dispatch='2*n_jobs',
       refit=True, return_train_score='warn', scoring=None, verbose=0)

In [268]:
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(C=11, penalty='l2').fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.4f}".format(clf.score(test_document, Y_test)))

Training set score: 1.00
Test set score: 0.8962


In [233]:
from sklearn.linear_model import SGDClassifier
clf = SGDClassifier().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.4f}".format(clf.score(test_document, Y_test)))

C:\Users\sunilchawla\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


Training set score: 0.96
Test set score: 0.8956


In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_vect = TfidfVectorizer(ngram_range=(1, 3),min_df=3,max_df=0.56)
train_document = tfidf_vect.fit_transform(X_train)
test_document = tfidf_vect.transform(X_test)

In [232]:
from sklearn.svm import LinearSVC
clf = LinearSVC(C=0.3).fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.4f}".format(clf.score(test_document, Y_test)))

Training set score: 0.99
Test set score: 0.8973


# <br><font color="blue">highest accuracy = 89.73% with hyperparameter tuning in TF-IDF and LinearSVC(after data preprocessing) </font></br>